flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,410 @@
+
+  override def open(config: Configuration) {
+    LOG.debug(s"Compiling JoinFunction: $genJoinFuncName \n\n " +
+      s"Code:\n$genJoinFuncCode")
+    val clazz = compile(
+      getRuntimeContext.getUserCodeClassLoader,
+      genJoinFuncName,
+      genJoinFuncCode)
+    LOG.debug("Instantiating JoinFunction.")
+    joinFunction = clazz.newInstance()
+
+    cRowWrapper = new CRowWrappingCollector()
+    cRowWrapper.setChange(true)
+
+    // Initialize the data caches.
+    val leftListTypeInfo: TypeInformation[JList[Row]] = new ListTypeInfo[Row](leftType)
+    val leftStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]](
+        "InnerJoinLeftCache",
+        Types.LONG.asInstanceOf[TypeInformation[Long]],
+        leftListTypeInfo)
+    leftCache = getRuntimeContext.getMapState(leftStateDescriptor)
+
+    val rightListTypeInfo: TypeInformation[JList[Row]] = new ListTypeInfo[Row](rightType)
+    val rightStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]](
+        "InnerJoinRightCache",
+        Types.LONG.asInstanceOf[TypeInformation[Long]],
+        rightListTypeInfo)
+    rightCache = getRuntimeContext.getMapState(rightStateDescriptor)
+
+    // Initialize the timer states.
+    val leftTimerStateDesc: ValueStateDescriptor[Long] =
+      new ValueStateDescriptor[Long]("InnerJoinLeftTimerState", classOf[Long])
+    leftTimerState = getRuntimeContext.getState(leftTimerStateDesc)
+
+    val rightTimerStateDesc: ValueStateDescriptor[Long] =
+      new ValueStateDescriptor[Long]("InnerJoinRightTimerState", classOf[Long])
+    rightTimerState = getRuntimeContext.getState(rightTimerStateDesc)
+  }
+
+  /**
+    * Process rows from the left stream.
+    */
+  override def processElement1(
+      cRowValue: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+    updateOperatorTime(ctx)
+    val leftRow = cRowValue.row
+    val timeForLeftRow: Long = getTimeForLeftStream(ctx, leftRow)
+    val rightQualifiedLowerBound: Long = timeForLeftRow - rightRelativeSize
+    val rightQualifiedUpperBound: Long = timeForLeftRow + leftRelativeSize
+    cRowWrapper.out = out
+
+    if (rightOperatorTime < rightQualifiedUpperBound) {
+      // Put the leftRow into the cache for later use.
+      var leftRowList = leftCache.get(timeForLeftRow)
+      if (null == leftRowList) {
+        leftRowList = new ArrayList[Row](1)
+      }
+      leftRowList.add(leftRow)
+      leftCache.put(timeForLeftRow, leftRowList)
+      if (rightTimerState.value == 0) {
+        // Register a timer on the RIGHT stream to remove rows.
+        registerCleanUpTimer(ctx, timeForLeftRow, rightTimerState, leftRow = true)
+      }
+    }
+    // We'd like to produce as many results as possible.
+    if (rightExpirationTime < rightQualifiedUpperBound) {
```

Comment:
Because it's the previous expiration time (either being calculated in `processElement()` or `onTime()`) that really decides the state of the current cache. Say `prevMinTime` and `prevExpirationTime` are the contained minimum time and the expiration time of the cache **before** processing the current row. The condition here should be `prevMinTime < rightQualifiedUpperBound`. However, since the `prevMinTime` is hard to be calculated, I use `prevExpirationTime` as an approximation of `prevMinTime` here. Actually,  `prevExpirationTime <= prevMinTime` is always `true`, thus we won't miss any opportunities to join.

Reply:
OK, that makes sense. Thanks for the explanation. Can you add a brief comment to explain the condition?